Firestore/core/src/firebase/firestore/local/leveldb_lru_reference_delegate.mm

```diff
@@ -0,0 +1,188 @@
+LevelDbLruReferenceDelegate::~LevelDbLruReferenceDelegate() = default;
+
+void LevelDbLruReferenceDelegate::Start() {
+  ListenSequenceNumber highest_sequence_number =
+      db_->query_cache()->highest_listen_sequence_number();
+  listen_sequence_ = absl::make_unique<ListenSequence>(highest_sequence_number);
+}
+
+void LevelDbLruReferenceDelegate::AddInMemoryPins(ReferenceSet* set) {
+  // We should be able to assert that additional_references_ is nullptr, but
+  // due to restarts in spec tests it would fail.
+  additional_references_ = set;
+}
+
+void LevelDbLruReferenceDelegate::AddReference(const DocumentKey& key) {
+  WriteSentinel(key);
+}
+
+void LevelDbLruReferenceDelegate::RemoveReference(const DocumentKey& key) {
+  WriteSentinel(key);
+}
+
+void LevelDbLruReferenceDelegate::RemoveMutationReference(
+    const DocumentKey& key) {
+  WriteSentinel(key);
+}
+
+void LevelDbLruReferenceDelegate::RemoveTarget(const QueryData& query_data) {
+  QueryData updated =
+      query_data.Copy(query_data.snapshot_version(), query_data.resume_token(),
+                      current_sequence_number());
+  db_->query_cache()->UpdateTarget(std::move(updated));
+}
+
+void LevelDbLruReferenceDelegate::UpdateLimboDocument(const DocumentKey& key) {
+  WriteSentinel(key);
+}
+
+ListenSequenceNumber LevelDbLruReferenceDelegate::current_sequence_number() {
+  HARD_ASSERT(current_sequence_number_ != kListenSequenceNumberInvalid,
+              "Asking for a sequence number outside of a transaction");
+  return current_sequence_number_;
+}
+
+void LevelDbLruReferenceDelegate::OnTransactionStarted(absl::string_view) {
+  HARD_ASSERT(current_sequence_number_ == kListenSequenceNumberInvalid,
+              "Previous sequence number is still in effect");
+  current_sequence_number_ = listen_sequence_->Next();
+}
+
+void LevelDbLruReferenceDelegate::OnTransactionCommitted() {
+  current_sequence_number_ = kListenSequenceNumberInvalid;
+}
+
+LruGarbageCollector* LevelDbLruReferenceDelegate::garbage_collector() {
+  return gc_.get();
+}
+
+int64_t LevelDbLruReferenceDelegate::CalculateByteSize() {
+  return db_->CalculateByteSize();
+}
+
+size_t LevelDbLruReferenceDelegate::GetSequenceNumberCount() {
+  size_t total_count = db_->query_cache()->size();
+  EnumerateOrphanedDocuments(
+      [&total_count](const DocumentKey&, ListenSequenceNumber) {
+        total_count++;
+      });
+  return total_count;
+}
+
+void LevelDbLruReferenceDelegate::EnumerateTargets(
+    const TargetCallback& callback) {
+  db_->query_cache()->EnumerateTargets(callback);
+}
+
+void LevelDbLruReferenceDelegate::EnumerateOrphanedDocuments(
+    const OrphanedDocumentCallback& callback) {
+  db_->query_cache()->EnumerateOrphanedDocuments(callback);
+}
+
+int LevelDbLruReferenceDelegate::RemoveOrphanedDocuments(
+    ListenSequenceNumber upper_bound) {
+  int count = 0;
+  db_->query_cache()->EnumerateOrphanedDocuments(
+      [&count, this, upper_bound](const DocumentKey& key,
```

Comment:
This is fine. You could optionally just use a default capture `[&]` (since the lifetime of the lambda is shorter than the function.)

Reply:
Done.